servant-server/src/Servant/Server/Internal/Router.hs

```diff
@@ -146,45 +136,46 @@ routerLayout router =
     mkSubTree False path children = ("└─ " <> path <> "/") : map ("   " <>) children
 
 -- | Apply a transformation to the response of a `Router`.
-tweakResponse :: (RouteResult Response -> RouteResult Response) -> Router -> Router
+tweakResponse :: (RouteResult Response -> RouteResult Response) -> Router env -> Router env
 tweakResponse f = fmap (\a -> \req cont -> a req (cont . f))
 
 -- | Interpret a router as an application.
-runRouter :: Router -> RoutingApplication
-runRouter (WithRequest router) request respond =
-  runRouter (router request) request respond
-runRouter (StaticRouter table ls) request respond =
+runRouter :: Router () -> RoutingApplication
+runRouter r = runRouterEnv r ()
+
+runRouterEnv :: Router env -> env -> RoutingApplication
+runRouterEnv (StaticRouter table ls) env request respond =
   case pathInfo request of
-    []   -> runChoice ls request respond
+    []   -> runChoice ls env request respond
     -- This case is to handle trailing slashes.
-    [""] -> runChoice ls request respond
+    [""] -> runChoice ls env request respond
     first : rest | Just router <- M.lookup first table
       -> let request' = request { pathInfo = rest }
-         in  runRouter router request' respond
+         in  runRouterEnv router env request' respond
     _ -> respond $ Fail err404
-runRouter (DynamicRouter fun)  request respond =
+runRouterEnv (DynamicRouter router)  env request respond =
   case pathInfo request of
     []   -> respond $ Fail err404
     -- This case is to handle trailing slashes.
     [""] -> respond $ Fail err404
     first : rest
       -> let request' = request { pathInfo = rest }
-         in  runRouter (fun first) request' respond
-runRouter (RawRouter app)      request respond = app request respond
-runRouter (Choice r1 r2)       request respond =
-  runChoice [runRouter r1, runRouter r2] request respond
+         in  runRouterEnv router (first, env) request' respond
+runRouterEnv (RawRouter app)         env request respond = app env request respond
+runRouterEnv (Choice r1 r2)          env request respond =
+  runChoice [runRouterEnv r1, runRouterEnv r2] env request respond
```

Comment:
I suspect this would be easier to read with a `case` statement for the first argument (`Router env`).